includes/class-edd-exception.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Exception object for EDD
+ *
+ * This class handles connecting exception handling to the debug log (as needed).
+ *
+ * @package     EDD
+ * @subpackage  Classes/Utilities
+ * @copyright   Copyright (c) 2018, Pippin Williamson
+ * @license     http://opensource.org/licenses/gpl-2.0.php GNU Public License
+ * @since       3.0
+ */
+
+/**
+ * Implements some EDD-specific logic for exception handling.
+ *
+ * @since 3.0
+ *
+ * @see \Exception
+ */
+class EDD_Exception extends \Exception {
+
+	/**
+	 * Logs an exception to the EDD debug log.
+	 *
+	 * @since 3.0
+	 *
+	 * @param string $extra Any extra information to include in the log entry
+	 *                      alongside the exception message.
+	 */
+	public function log( $extra = '' ) {
+		/**
+		 * Filters whether to skip logging EDD exceptions to the debug log.
+		 *
+		 * @since 3.0
+		 *
+		 * @param bool $skip Whether to skip logging exceptions. Default false (proceed).
+		 */
+		if ( false === apply_filters( 'edd_skip_logging_exceptions', false ) ) {
```

Comment:
@DrewAPicture I'm wondering if this filter could cause confusion or problems. Example: say I'm using/extending this class and I always want to log exceptions. Other code could break my exception logging, and anyone else's exception logging. Errors could possibly disappear into a black hole. Is there a solid use case for disabling exception logging when all other debug logging is enabled?

Reply:
By allowing exceptions to be logged to the debug log, there's potential to _quickly_ fill up logs with said exceptions if there's a problem. I think in any situation where many pieces of code can have their hands in the same pot there are going to be cases where one preempts the other and that's a reality we have to be OK with, especially in something as yuge as EDD.

Reply:
Large log files could definitely happen, but EDD debug mode is meant for debugging, not full-time production use. If support asks customers to enable debug logging and exceptions aren't logged, we're missing some important info that quickly solve problems.

Reply:
The filter exists only for flexibility. which EDD _inherently_ needs in order to exist in the ecosystem it does. I'd be fine with removing it.